# 02. Analysis Seoul Crime

#### 서울 강남3구 체감 안전도 높다는 기사를 봤다
#### 실제 강남3구가 범죄로부터 안전하다고 말할 수 있는지 확인해보자

### +GoogleMaps, Folium, Seaborn, Pandas의 Pivot_table 익히기

In [4]:
import numpy as np
import pandas as pd

- 숫자값들이 콤마(,)를 사용하고 있어서 문자로 인식될 수 있다.
- 천단위 구분(thousands = ',')이라고 알려주면 콤마를 제거하고 숫자형으로 읽는다

In [6]:
crime_raw_data = pd.read_csv('../data/02. crime_in_Seoul.csv', thousands = ",",
                            encoding = "euc-kr")
crime_raw_data.head(3)

구분  죄종 발생검거   건수
0  중부  살인   발생  2.0
1  중부  살인   검거  2.0
2  중부  강도   발생  3.0

In [7]:
crime_raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65534 entries, 0 to 65533
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   구분      310 non-null    object 
 1   죄종      310 non-null    object 
 2   발생검거    310 non-null    object 
 3   건수      310 non-null    float64
dtypes: float64(1), object(3)
memory usage: 2.0+ MB


- 위의 결과를 보면 index는 65533인데, 데이터가 310개?
- 특정 column에서 unique 검사해보기

In [8]:
crime_raw_data['죄종'].unique()  #nan이 있다.

array(['살인', '강도', '강간', '절도', '폭력', nan], dtype=object)

In [9]:
crime_raw_data[crime_raw_data['죄종'].isnull()]

구분   죄종 발생검거  건수
310    NaN  NaN  NaN NaN
311    NaN  NaN  NaN NaN
312    NaN  NaN  NaN NaN
313    NaN  NaN  NaN NaN
314    NaN  NaN  NaN NaN
...    ...  ...  ...  ..
65529  NaN  NaN  NaN NaN
65530  NaN  NaN  NaN NaN
65531  NaN  NaN  NaN NaN
65532  NaN  NaN  NaN NaN
65533  NaN  NaN  NaN NaN

[65224 rows x 4 columns]

- 황당하게도 실제 데이터(엑셀에서 읽을 때는 정상이지만) 그 후 엄청 많은 nan데이터 발생
- 이는 index가 65535의 크기를 가지게 되면서 실제 value와의 크기 차이가 발생했기 떄문이다.
- 이럴 때는 nan을 제거하는 것이 아니라 nan이 아닌 데이터만 다시 가져오자
- 이 상황은 연도별로 다르다

In [10]:
crime_raw_data = crime_raw_data[crime_raw_data['죄종'].notnull()]

In [12]:
crime_raw_data.info() #memory가 2MB-> 12.1KB로 엄청 줄었다.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 310 entries, 0 to 309
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   구분      310 non-null    object 
 1   죄종      310 non-null    object 
 2   발생검거    310 non-null    object 
 3   건수      310 non-null    float64
dtypes: float64(1), object(3)
memory usage: 12.1+ KB


### 단순 나열된 데이터를 원하는 데이터 열로 재배치 할 필요가 있다. 

In [15]:
crime_raw_data.head()

구분  죄종 발생검거     건수
0  중부  살인   발생    2.0
1  중부  살인   검거    2.0
2  중부  강도   발생    3.0
3  중부  강도   검거    3.0
4  중부  강간   발생  141.0

# Pandas의 Pivot_table

- 간단한 판매 현황표

In [19]:
df = pd.read_excel("../data/02. sales-funnel.xlsx")
df.head() 

Account                          Name            Rep       Manager  \
0   714466               Trantow-Barrows   Craig Booker  Debra Henley   
1   714466               Trantow-Barrows   Craig Booker  Debra Henley   
2   714466               Trantow-Barrows   Craig Booker  Debra Henley   
3   737550  Fritsch, Russel and Anderson   Craig Booker  Debra Henley   
4   146832                  Kiehn-Spinka  Daniel Hilton  Debra Henley   

       Product  Quantity  Price     Status  
0          CPU         1  30000  presented  
1     Software         1  10000  presented  
2  Maintenance         2   5000    pending  
3          CPU         1  35000   declined  
4          CPU         2  65000        won

In [20]:
pd.pivot_table(df, index = ['Name'])

Account   Price  Quantity
Name                                                   
Barton LLC                     740150   35000  1.000000
Fritsch, Russel and Anderson   737550   35000  1.000000
Herman LLC                     141962   65000  2.000000
Jerde-Hilpert                  412290    5000  2.000000
Kassulke, Ondricka and Metz    307599    7000  3.000000
Keeling LLC                    688981  100000  5.000000
Kiehn-Spinka                   146832   65000  2.000000
Koepp Ltd                      729833   35000  2.000000
Kulas Inc                      218895   25000  1.500000
Purdy-Kunde                    163416   30000  1.000000
Stokes LLC                     239344    7500  1.000000
Trantow-Barrows                714466   15000  1.333333

In [21]:
pd.pivot_table(df, index = ['Name', "Rep", "Manager"])

Account   Price  \
Name                         Rep           Manager                          
Barton LLC                   John Smith    Debra Henley    740150   35000   
Fritsch, Russel and Anderson Craig Booker  Debra Henley    737550   35000   
Herman LLC                   Cedric Moss   Fred Anderson   141962   65000   
Jerde-Hilpert                John Smith    Debra Henley    412290    5000   
Kassulke, Ondricka and Metz  Wendy Yule    Fred Anderson   307599    7000   
Keeling LLC                  Wendy Yule    Fred Anderson   688981  100000   
Kiehn-Spinka                 Daniel Hilton Debra Henley    146832   65000   
Koepp Ltd                    Wendy Yule    Fred Anderson   729833   35000   
Kulas Inc                    Daniel Hilton Debra Henley    218895   25000   
Purdy-Kunde                  Cedric Moss   Fred Anderson   163416   30000   
Stokes LLC                   Cedric Moss   Fred Anderson   239344    7500   
Trantow-Barrows              Craig Booker  Debra Henley    714466   15000   

                                                          Quantity  
Name                         Rep           Manager                  
Barton LLC                   John Smith    Debra Henley   1.000000  
Fritsch, Russel and Anderson Craig Booker  Debra Henley   1.000000  
Herman LLC                   Cedric Moss   Fred Anderson  2.000000  
Jerde-Hilpert                John Smith    Debra Henley   2.000000  
Kassulke, Ondricka and Metz  Wendy Yule    Fred Anderson  3.000000  
Keeling LLC                  Wendy Yule    Fred Anderson  5.000000  
Kiehn-Spinka                 Daniel Hilton Debra Henley   2.000000  
Koepp Ltd                    Wendy Yule    Fred Anderson  2.000000  
Kulas Inc                    Daniel Hilton Debra Henley   1.500000  
Purdy-Kunde                  Cedric Moss   Fred Anderson  1.000000  
Stokes LLC                   Cedric Moss   Fred Anderson  1.000000  
Trantow-Barrows              Craig Booker  Debra Henley   1.333333

In [23]:
pd.pivot_table(df, index = ['Name', "Rep"], values = ['Price']) #default는 평균이다.

Price
Name                         Rep                  
Barton LLC                   John Smith      35000
Fritsch, Russel and Anderson Craig Booker    35000
Herman LLC                   Cedric Moss     65000
Jerde-Hilpert                John Smith       5000
Kassulke, Ondricka and Metz  Wendy Yule       7000
Keeling LLC                  Wendy Yule     100000
Kiehn-Spinka                 Daniel Hilton   65000
Koepp Ltd                    Wendy Yule      35000
Kulas Inc                    Daniel Hilton   25000
Purdy-Kunde                  Cedric Moss     30000
Stokes LLC                   Cedric Moss      7500
Trantow-Barrows              Craig Booker    15000

In [24]:
pd.pivot_table(df, index = ['Name', "Rep"], values = ['Price'], aggfunc = np.sum)

Price
Name                         Rep                  
Barton LLC                   John Smith      35000
Fritsch, Russel and Anderson Craig Booker    35000
Herman LLC                   Cedric Moss     65000
Jerde-Hilpert                John Smith       5000
Kassulke, Ondricka and Metz  Wendy Yule       7000
Keeling LLC                  Wendy Yule     100000
Kiehn-Spinka                 Daniel Hilton   65000
Koepp Ltd                    Wendy Yule      70000
Kulas Inc                    Daniel Hilton   50000
Purdy-Kunde                  Cedric Moss     30000
Stokes LLC                   Cedric Moss     15000
Trantow-Barrows              Craig Booker    45000

In [26]:
pd.pivot_table(df, index = ['Name', "Rep"], values = ['Price'], aggfunc = [np.sum, len])

sum   len
                                             Price Price
Name                         Rep                        
Barton LLC                   John Smith      35000     1
Fritsch, Russel and Anderson Craig Booker    35000     1
Herman LLC                   Cedric Moss     65000     1
Jerde-Hilpert                John Smith       5000     1
Kassulke, Ondricka and Metz  Wendy Yule       7000     1
Keeling LLC                  Wendy Yule     100000     1
Kiehn-Spinka                 Daniel Hilton   65000     1
Koepp Ltd                    Wendy Yule      70000     2
Kulas Inc                    Daniel Hilton   50000     2
Purdy-Kunde                  Cedric Moss     30000     1
Stokes LLC                   Cedric Moss     15000     2
Trantow-Barrows              Craig Booker    45000     3

In [28]:
pd.pivot_table(
    df,
    index = ["Manager", "Rep"],
    values = ["Price"],
    columns = ["Product"],
    aggfunc = [np.sum],
    fill_value = 0,  #NaN에 대한 처리 지정
    )

sum                             
                              Price                             
Product                         CPU Maintenance Monitor Software
Manager       Rep                                               
Debra Henley  Craig Booker    65000        5000       0    10000
              Daniel Hilton  105000           0       0    10000
              John Smith      35000        5000       0        0
Fred Anderson Cedric Moss     95000        5000       0    10000
              Wendy Yule     165000        7000    5000        0

In [29]:
pd.pivot_table(
    df,
    index = ["Manager", "Rep", "Product"],
    values = ["Price","Quantity"],
    aggfunc = [np.sum],
    fill_value = 0,  #NaN에 대한 처리 지정
    )

sum         
                                          Price Quantity
Manager       Rep           Product                     
Debra Henley  Craig Booker  CPU           65000        2
                            Maintenance    5000        2
                            Software      10000        1
              Daniel Hilton CPU          105000        4
                            Software      10000        1
              John Smith    CPU           35000        1
                            Maintenance    5000        2
Fred Anderson Cedric Moss   CPU           95000        3
                            Maintenance    5000        1
                            Software      10000        1
              Wendy Yule    CPU          165000        7
                            Maintenance    7000        3
                            Monitor        5000        2

In [34]:
pd.pivot_table(
    df,
    index = ["Manager", "Rep", "Product"],
    values = ["Price","Quantity"],
    aggfunc = [np.sum, np.mean],
    fill_value = 0,  #NaN에 대한 처리 지정
    margins = True, #총합계 나타내기
    )

sum                   mean  \
                                          Price Quantity         Price   
Manager       Rep           Product                                      
Debra Henley  Craig Booker  CPU           65000        2  32500.000000   
                            Maintenance    5000        2   5000.000000   
                            Software      10000        1  10000.000000   
              Daniel Hilton CPU          105000        4  52500.000000   
                            Software      10000        1  10000.000000   
              John Smith    CPU           35000        1  35000.000000   
                            Maintenance    5000        2   5000.000000   
Fred Anderson Cedric Moss   CPU           95000        3  47500.000000   
                            Maintenance    5000        1   5000.000000   
                            Software      10000        1  10000.000000   
              Wendy Yule    CPU          165000        7  82500.000000   
                            Maintenance    7000        3   7000.000000   
                            Monitor        5000        2   5000.000000   
All                                      522000       30  30705.882353   

                                                   
                                         Quantity  
Manager       Rep           Product                
Debra Henley  Craig Booker  CPU          1.000000  
                            Maintenance  2.000000  
                            Software     1.000000  
              Daniel Hilton CPU          2.000000  
                            Software     1.000000  
              John Smith    CPU          1.000000  
                            Maintenance  2.000000  
Fred Anderson Cedric Moss   CPU          1.500000  
                            Maintenance  1.000000  
                            Software     1.000000  
              Wendy Yule    CPU          3.500000  
                            Maintenance  3.000000  
                            Monitor      2.000000  
All                                      1.764706

# 경찰서 Data를 Pivot table로 나타내보기

In [35]:
crime_raw_data.head()

구분  죄종 발생검거     건수
0  중부  살인   발생    2.0
1  중부  살인   검거    2.0
2  중부  강도   발생    3.0
3  중부  강도   검거    3.0
4  중부  강간   발생  141.0

- 경찰서 이름을 index로 하자
- default가 평균(mean)이므로 사건의 합을 기록하기 위해 aggfunc 옵션에 sum 사용하는 것 주의!

In [38]:
crime_station = crime_raw_data.pivot_table(
    crime_raw_data,
    index = ["구분"],
    columns = ["죄종", "발생검거"],
    aggfunc = [np.sum])
crime_station.head()

sum                                                              
         건수                                                              
죄종       강간           강도          살인           절도              폭력        
발생검거     검거     발생    검거    발생    검거   발생      검거      발생      검거      발생
구분                                                                       
강남    269.0  339.0  26.0  24.0   3.0  3.0  1129.0  2438.0  2096.0  2336.0
강동    152.0  160.0  13.0  14.0   5.0  4.0   902.0  1754.0  2201.0  2530.0
강북    159.0  217.0   4.0   5.0   6.0  7.0   672.0  1222.0  2482.0  2778.0
강서    239.0  275.0  10.0  10.0  10.0  9.0  1070.0  1952.0  2768.0  3204.0
관악    264.0  322.0  10.0  12.0   7.0  6.0   937.0  2103.0  2707.0  3235.0

#### 깔끔하게 정리됐지만, 이런 경우 column이 multi로 잡힌다는 것이 함정이다.

### multi columns index 발생한다.
- pivot_table을 적용하면 column이나 index가 다중으로 잡힌다.

In [41]:
crime_station.columns 

MultiIndex([('sum', '건수', '강간', '검거'),
            ('sum', '건수', '강간', '발생'),
            ('sum', '건수', '강도', '검거'),
            ('sum', '건수', '강도', '발생'),
            ('sum', '건수', '살인', '검거'),
            ('sum', '건수', '살인', '발생'),
            ('sum', '건수', '절도', '검거'),
            ('sum', '건수', '절도', '발생'),
            ('sum', '건수', '폭력', '검거'),
            ('sum', '건수', '폭력', '발생')],
           names=[None, None, '죄종', '발생검거'])

#### Multi Index에 대한 접근

In [42]:
crime_station["sum", "건수", "강도", "검거"].head()

구분
강남    26.0
강동    13.0
강북     4.0
강서    10.0
관악    10.0
Name: (sum, 건수, 강도, 검거), dtype: float64

#### 다중 컬럼에서 특정 컬럼 제거

In [43]:
crime_station.columns = crime_station.columns.droplevel([0,1])
crime_station.columns

MultiIndex([('강간', '검거'),
            ('강간', '발생'),
            ('강도', '검거'),
            ('강도', '발생'),
            ('살인', '검거'),
            ('살인', '발생'),
            ('절도', '검거'),
            ('절도', '발생'),
            ('폭력', '검거'),
            ('폭력', '발생')],
           names=['죄종', '발생검거'])

In [44]:
crime_station.head()

죄종       강간           강도          살인           절도              폭력        
발생검거     검거     발생    검거    발생    검거   발생      검거      발생      검거      발생
구분                                                                       
강남    269.0  339.0  26.0  24.0   3.0  3.0  1129.0  2438.0  2096.0  2336.0
강동    152.0  160.0  13.0  14.0   5.0  4.0   902.0  1754.0  2201.0  2530.0
강북    159.0  217.0   4.0   5.0   6.0  7.0   672.0  1222.0  2482.0  2778.0
강서    239.0  275.0  10.0  10.0  10.0  9.0  1070.0  1952.0  2768.0  3204.0
관악    264.0  322.0  10.0  12.0   7.0  6.0   937.0  2103.0  2707.0  3235.0

#### 현재 index는 경찰서 이름으로 되어있다 --> 경찰서 이름으로 구 이름을 알아야 한다.

In [45]:
crime_station.index

Index(['강남', '강동', '강북', '강서', '관악', '광진', '구로', '금천', '남대문', '노원', '도봉',
       '동대문', '동작', '마포', '방배', '서대문', '서부', '서초', '성동', '성북', '송파', '수서',
       '양천', '영등포', '용산', '은평', '종로', '종암', '중랑', '중부', '혜화'],
      dtype='object', name='구분')

In [50]:
# !pip list
#get_ipython().system("pip list")

# Google Maps API 설치

In [1]:
!pip install googlemaps

  Stored in directory: C:\Users\whfh1\AppData\Local\pip\Cache\wheels\e7\db\40\c2eb6898ee3e70fe18cbaf63f4168b480be62a5de8feae6f1e
Successfully built googlemaps


In [2]:
import googlemaps

In [3]:
gmaps_key = "AIzaSyAl_zjhfKXa2bKwvuMKhNHxfv9QC9zBwqk"
gmaps = googlemaps.Client(key=gmaps_key)

In [5]:
gmaps.geocode("서울영등포경찰서", language = "ko")

[{'address_components': [{'long_name': '６１８−７',
    'short_name': '６１８−７',
    'types': ['premise']},
   {'long_name': '영등포동1가',
    'short_name': '영등포동1가',
    'types': ['political', 'sublocality', 'sublocality_level_2']},
   {'long_name': '영등포구',
    'short_name': '영등포구',
    'types': ['political', 'sublocality', 'sublocality_level_1']},
   {'long_name': '서울특별시',
    'short_name': '서울특별시',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': '대한민국',
    'short_name': 'KR',
    'types': ['country', 'political']},
   {'long_name': '150-031',
    'short_name': '150-031',
    'types': ['postal_code']}],
  'formatted_address': '대한민국 서울특별시 영등포구 영등포동1가 618-7',
  'geometry': {'location': {'lat': 37.5153176, 'lng': 126.905728},
   'location_type': 'ROOFTOP',
   'viewport': {'northeast': {'lat': 37.51666658029149,
     'lng': 126.9070769802915},
    'southwest': {'lat': 37.51396861970849, 'lng': 126.9043790197085}}},
  'place_id': 'ChIJhaBTdfyefDURZO91m0CtbuM',
  'plus_

In [ ]:
# Pandas에 잘 맞춰진 반복문용 명령